src/NovelRT/Experimental/Graphics/Vulkan/VulkanGraphicsTexture.cpp

```diff
@@ -0,0 +1,348 @@
+        }
+
+        return reinterpret_cast<uint8_t*>(pDestination) + readRangeOffset;
+    }
+
+    void VulkanGraphicsTexture::Unmap()
+    {
+        std::shared_ptr<VulkanGraphicsDevice> device = GetAllocator()->GetDevice();
+
+        VkDevice vulkanDevice = device->GetVulkanDevice();
+        VkDeviceMemory vulkanDeviceMemory = GetBlock()->GetVulkanDeviceMemory();
+
+        vkUnmapMemory(vulkanDevice, vulkanDeviceMemory);
+    }
+
+    void VulkanGraphicsTexture::UnmapAndWrite()
+    {
+        std::shared_ptr<VulkanGraphicsDevice> device = GetAllocator()->GetDevice();
+
+        VkDevice vulkanDevice = device->GetVulkanDevice();
+        VkDeviceMemory vulkanDeviceMemory = GetBlock()->GetVulkanDeviceMemory();
+
+        size_t nonCoherentAtomSize =
+            device->GetAdapter()->GetVulkanPhysicalDeviceProperties().limits.nonCoherentAtomSize;
+        size_t offset = GetOffset();
+        size_t size = (GetSize() + nonCoherentAtomSize - 1) & ~(nonCoherentAtomSize - 1);
+
+        VkMappedMemoryRange mappedMemoryRange{};
+        mappedMemoryRange.sType = VK_STRUCTURE_TYPE_MAPPED_MEMORY_RANGE;
+        mappedMemoryRange.memory = vulkanDeviceMemory;
+        mappedMemoryRange.offset = offset;
+        mappedMemoryRange.size = size;
+
+        VkResult flushResult = vkFlushMappedMemoryRanges(vulkanDevice, 1, &mappedMemoryRange);
+
+        if (flushResult != VK_SUCCESS)
+        {
+            // TODO: Add real exception.
+            throw std::runtime_error("Failed to flush memory ranges after writing! Reason: " +
+                                     std::to_string(flushResult));
+        }
+
+        vkUnmapMemory(vulkanDevice, vulkanDeviceMemory);
+    }
+
+    void VulkanGraphicsTexture::UnmapAndWrite(size_t writtenRangeOffset, size_t writtenRangeLength)
+    {
+        std::shared_ptr<VulkanGraphicsDevice> device = GetAllocator()->GetDevice();
+
+        VkDevice vulkanDevice = device->GetVulkanDevice();
+        VkDeviceMemory vulkanDeviceMemory = GetBlock()->GetVulkanDeviceMemory();
+
+        if (writtenRangeLength != 0)
+        {
+            size_t nonCoherentAtomSize =
+                device->GetAdapter()->GetVulkanPhysicalDeviceProperties().limits.nonCoherentAtomSize;
+            size_t offset = GetOffset() + writtenRangeOffset;
+            size_t size = (writtenRangeLength + nonCoherentAtomSize - 1) & ~(nonCoherentAtomSize - 1);
+
+            VkMappedMemoryRange mappedMemoryRange{};
+            mappedMemoryRange.sType = VK_STRUCTURE_TYPE_MAPPED_MEMORY_RANGE;
+            mappedMemoryRange.memory = vulkanDeviceMemory;
+            mappedMemoryRange.offset = offset;
+            mappedMemoryRange.size = size;
+
+            VkResult flushResult = vkFlushMappedMemoryRanges(vulkanDevice, 1, &mappedMemoryRange);
+
+            if (flushResult != VK_SUCCESS)
+            {
+                // TODO: Add real exception.
+                throw std::runtime_error("Failed to flush memory ranges after writing! Reason: " +
```

Comment:
;_; Add a real exception pls?

Reply:
This will probably be another design ticket or part of an existing one that I mentioned previously.